### PR TITLE
SetupWrapper: separate determining and running Setup

### PR DESCRIPTION
**Pull request by ttuegel**
It is useful to know outside of `setupWrapper` which Setup version will
be used, so the two phases (knowing and running) are separated.

This is merely refactoring. I need it for my `reconfigure` branch and I don't want it to get merge-conflicted again while I finish that feature.

### CODE REVIEW

**Comment by ezyang** [comment]
So, if I understand correctly, the refactor introduces two new types `Setup` and `SetupMethod` which capture the intermediate result of determining what Setup will be used, but otherwise keeps the logic the same? If so, sounds good to me; but let's document the new types.  (The commit is a bit fuzzy so it was hard to follow.)

**Comment by ttuegel** [comment]
Yeah, I still have to fix some merge conflicts that I resolved incorrectly.

Is it a known bug that test output doesn't show up in AppVeyor?

**Comment by ezyang** [comment]
No; I've never seen the test suite do that log before.

**Comment by 23Skidoo** [comment]
Refactoring SetupWrapper is a good idea in general.

**Comment by dcoutts** [comment]
@ttuegel I've not looked at the code yet, but I support the principle. Is there any chance that having separated out the working of which Cabal CLI/API version the Setup supports from calling it, that it becomes possible to skip the first step entirely if we know in advance what version will be used? I ask because the "new-build" code path does know up front what Cabal CLI version will be used and this could help simplify things a bit.

**Comment by grayjay** [comment]
> @ttuegel I've not looked at the code yet, but I support the principle. Is there any chance that having separated out the working of which Cabal CLI/API version the Setup supports from calling it, that it becomes possible to skip the first step entirely if we know in advance what version will be used? I ask because the "new-build" code path does know up front what Cabal CLI version will be used and this could help simplify things a bit.

#3723 is related.

**Comment by ttuegel** [comment]
> I've not looked at the code yet, but I support the principle. Is there any chance that having separated out the working of which Cabal CLI/API version the Setup supports from calling it, that it becomes possible to skip the first step entirely if we know in advance what version will be used? I ask because the "new-build" code path does know up front what Cabal CLI version will be used and this could help simplify things a bit.

If we know what Cabal version is being used and we know if we need an external, internal, or self-exec method, then we can simply construct a `Setup` with the correct values and skip the first step.
